modules/connectivity/ParticipantConnectionStatus.js

```diff
@@ -380,18 +455,117 @@ export default class ParticipantConnectionStatus {
             isConnActiveByJvb = true;
         }
 
-        const isConnectionActive
-            = isConnActiveByJvb
-                && (isVideoMuted || (isInLastN && !isVideoTrackFrozen));
+        let newState = ParticipantConnectionStatus.INACTIVE;
+
+        if (isConnActiveByJvb) {
+            if (isInLastN) {
+                if (isVideoTrackFrozen) {
+                    newState = this._isRestoringTimedout(id)
+                        ? ParticipantConnectionStatus.INTERRUPTED
+                            : ParticipantConnectionStatus.RESTORING;
+                } else {
+                    newState = ParticipantConnectionStatus.ACTIVE;
+                }
+            }
+        } else {
+            // when there is a connection problem signaled from jvb
+            // it means no media was flowing for at least 15secs, so everything
+            // should be interrupted, when in p2p mode we will never end up here
+            newState = ParticipantConnectionStatus.INTERRUPTED;
+        }
+
+        // if the new state is not restoring clear timers and timestamps
+        // that we use to track the restoring state
+        if (newState !== ParticipantConnectionStatus.RESTORING) {
+            this._clearRestoringTimer(id);
+        }
 
         logger.debug(
-            `Figure out conn status, is video muted: ${isVideoMuted
+            `Figure out conn status for ${id}, is video muted: ${isVideoMuted
                  } is active(jvb): ${isConnActiveByJvb
                  } video track frozen: ${isVideoTrackFrozen
                  } is in last N: ${isInLastN
-                 } => ${isConnectionActive}`);
+                 } current2NewStatus: ${participant.getConnectionStatus()
+                 } => ${newState}`);
+
+        this._changeConnectionStatus(participant, newState);
+    }
+
+    /**
+     * On change in Last N set check all leaving and entering participants to
+     * change their corresponding statuses.
+     *
+     * @param {Array<string>} leavingLastN array of ids leaving lastN.
+     * @param {Array<string>} enteringLastN array of ids entering lastN.
+     * @private
+     */
+    _onLastNChanged(leavingLastN, enteringLastN) {
+        if (leavingLastN) {
```

Comment:
use default parameters and a for..of loop and you'll save a level of indentation here